Rework Book (#1230)
graphql-rust/juniper: book/src/types/enums.md

Enums
=====

> [GraphQL enum][0] types, like [scalar][1] types, also represent leaf values in a GraphQL type system. However [enum][0] types describe the set of possible values.
>
> [Enums][0] are not references for a numeric value, but are unique values in their own right. They may serialize as a string: the name of the represented value.

> [GraphQL enum][0] in [Juniper] could be defined by using a [`#[derive(GraphQLEnum)]`][2] attribute on a [Rust enum][3] (its variants cannot have any fields, though):

Reviewer: ```suggestion
With [Juniper] a [GraphQL enum][0] may be defined by using the [`#[derive(GraphQLEnum)]`][2] attribute on a [Rust enum][3] as long as its variants do not have any fields:
```

```rust
# extern crate juniper;
# use juniper::GraphQLEnum;
#
/// This doc comment is visible only in Rust API docs.
#[derive(GraphQLEnum)]
#[graphql(description = "An episode of Star Wars")]
enum StarWarsEpisode {
    /// This doc comment is visible only in Rust API docs.
    #[graphql(description = "This description is visible only in GraphQL schema.")]
    NewHope,

    /// This doc comment is visible only in Rust API docs.
    #[graphql(desc = "Arguably the best one in the trilogy.")]
    //        ^^^^ shortcut for a `description` argument
    Empire,

    /// This doc comment is visible in both Rust API docs and GraphQL schema 
    /// descriptions.
    Jedi,
    
    #[deprecated(note = "Only visible in Rust.")]
    #[graphql(deprecated = "We don't really talk about this one.")]
    //        ^^^^^^^^^^ takes precedence over Rust's `#[deprecated]` attribute
    ThePhantomMenace, // has no description in GraphQL schema
}
#
# fn main() {}
```

> > **NOTE**: Only [GraphQL object][6]/[interface][7] fields and [GraphQL enum][0] values could be [deprecated][5].

Reviewer: ```suggestion
> **NOTE**: Only [GraphQL object][6]/[interface][7] fields and [GraphQL enum][0] values may be [deprecated][5].
```

### Ignoring

> By default, all [enum][3] variants are included into the generated [GraphQL enum][0] type as values. To prevent including a specific variant, it should be annotated with the `#[graphql(ignore)]` attribute:

Reviewer: ```suggestion
By default, all [enum][3] variants are included in the generated [GraphQL enum][0] type as values. To prevent including a specific variant, annotate it with the `#[graphql(ignore)]` attribute:
```

```rust
# #![allow(dead_code)]
# extern crate juniper;
# use juniper::GraphQLEnum;
#
#[derive(GraphQLEnum)]
enum Episode<T> {
    NewHope,
    Empire,
    Jedi,
    #[graphql(ignore)]
    Legends(T),   // cannot be queried from GraphQL
    #[graphql(skip)]
    //        ^^^^ alternative naming, up to your preference
    CloneWars(T), // cannot be queried from GraphQL
}
#
# fn main() {}
```

> > **TIP**: See more features in API docs of the [`#[derive(GraphQLEnum)]`][2] attribute.

Reviewer: ```suggestion
> **TIP**: See more available features in the API docs of the [`#[derive(GraphQLEnum)]`][2] attribute.
```
